delta-app/src/main/java/io/cdap/delta/app/service/AssessmentHandler.java

```diff
@@ -288,6 +285,88 @@ public void getState(HttpServiceRequest request, HttpServiceResponder responder,
+  @PUT
+  @Path("v1/contexts/{namespace}/apps/{app}/generations/{generation}/instanceIds/{instanceid}/offset")
+  public void saveOffsetData(HttpServiceRequest request, HttpServiceResponder responder,
+                             @PathParam("namespace") String namespaceName,
+                             @PathParam("app") String app,
+                             @PathParam("generation") long generation,
+                             @PathParam("instanceid") int instanceid) {
+    respond(namespaceName, responder, (namespace) -> {
+      DeltaWorkerId id = new DeltaWorkerId(new DeltaPipelineId(namespaceName, app, generation), instanceid);
+      String body = StandardCharsets.UTF_8.decode(request.getContent()).toString();
+      getStateStore().writeOffset(id, GSON.fromJson(body, OffsetAndSequence.class));
+      responder.sendStatus(HttpURLConnection.HTTP_OK);
+    });
+  }
+
+  //Looks up the delta_state_store table for the given state
+  @GET
+  @Path("v1/contexts/{namespace}/apps/{app}/generations/{generation}/instanceIds/{instanceid}/states/{state}")
+  public void getStateData(HttpServiceRequest request, HttpServiceResponder responder,
+                        @PathParam("namespace") String namespaceName,
+                        @PathParam("app") String app,
+                        @PathParam("generation") long generation,
+                        @PathParam("instanceid") int instanceid,
+                        @PathParam("state") String state) {
+    respond(namespaceName, responder, (namespace) -> {
+      DeltaWorkerId id = new DeltaWorkerId(new DeltaPipelineId(namespaceName, app, generation), instanceid);
+      byte[] data = getStateStore().readState(id, state);
+      responder.sendString(new String(data, StandardCharsets.UTF_8));
+    });
+  }
+
+  // Upserts in the delta_state_store table for state data.
+  @PUT
+  @Path("v1/contexts/{namespace}/apps/{app}/generations/{generation}/instanceIds/{instanceid}/states/{state}")
+  public void saveStateData(HttpServiceRequest request, HttpServiceResponder responder,
+                         @PathParam("namespace") String namespaceName,
+                         @PathParam("app") String app,
+                         @PathParam("generation") long generation,
+                         @PathParam("instanceid") int instanceid,
+                         @PathParam("state") String state) {
+
+    respond(namespaceName, responder, (namespace) -> {
+      DeltaWorkerId id = new DeltaWorkerId(new DeltaPipelineId(namespaceName, app, generation), instanceid);
+      ByteBuffer byteBuffer = request.getContent();
+      byte[] data = new byte[byteBuffer.remaining()];
+      byteBuffer.get(data);
+
+      getStateStore().writeState(id, state, data);
+      responder.sendStatus(HttpURLConnection.HTTP_OK);
+    });
+  }
+
+  //Fetches the latest generation for the given namespace and app.
+  @GET
+  @Path("v1/context/{namespace}/apps/{app}/maxgeneration")
```

Comment:
this seems like it shouldn't be exposed. Where is it being used?

Reply:
I see, it's in the StateStore interface. The Worker doesn't actually ever use it though does it? Looks like the remote implementation just returns null.

Reply:
This is used in Unit test cases [ for the main delta pipeline tests ].
Without these, there is no other way to fetch the max generation from Unit test case, which is necessary to create proper `DeltaWorkerId`